Full `file://` syntax support

feltech opened the pull request:
Closes #14. Fully support `file` URLs by making use the OpenAssetIO `FileUrlPathConverter` utility.

Also do a little refactoring to pimpl. 

Additionally, we seemed to have lost the revert commit that removed the mitigation of https://github.com/OpenAssetIO/OpenAssetIO/issues/1225, so update the exception catching to once again catch `std::exception`.

elliotcmorris commented on src/resolver.h:
```diff
@@ -22,40 +15,35 @@ class UsdOpenAssetIOResolver final : public PXR_NS::ArDefaultResolver {
-
-  /// Fixed OpenAssetIO context.
-  openassetio::ContextConstPtr context_;
+  class Impl;
+  std::unique_ptr<Impl> impl_;
```
[Consider] You spoke to me about how this was not neccesary. If there is no need for pimpl, why do it? 

It's your code at the end of the day, but all this accomplishes to my mind is :
- Create a much larger changeset than necessary
- Force a redundant heap allocation
- Mislead people into thinking there is a need to pimpl for interface stability reasons
- Make the code harder to understand for people who haven't encountered pimpl before

If your goal was to be "clean", ie, separating the deferment of failed results to the default resolver from the main implementation. That could be accomplished in a more direct way with free functions.

feltech replied:
I did roll back the pimpl, but then realised I'd end up `#include`ing  `FileUrlPathConverter` to add it as a new member, whereas other types were forward-declared in the header. A minor issue for sure, but it made me reflect on the style and I ended up liking the pimpl more, since it keeps the header completely clean, with no reference to OpenAssetIO at all.

I wonder now though if there is any reason to have a header at all, perhaps that's the real misleading thing...

feltech replied:
OK, completely refactored and force-pushed so there is no confusing interface header. Also split things up a bit better commit-wise

elliotcmorris commented on README.md:
```diff
@@ -81,13 +81,15 @@ limitations.
 
-Currently, OpenAssetIO must be built from source to build
-`usdOpenAssetIOResolver`.
+The OpenAssetIO libary is available pre-built for some common platforms
+are available
```
[Fix]
```suggestion
The OpenAssetIO libary is available pre-built for some common platforms.
These are available
```

feltech replied:
Good spot. Rolled into refactor - f939e714

feltech committed:
[Core] Remove exception RTTI bug mitigation

Part of #14. Fix `catchAndLogExceptions` to remove the mitigation
against bad RTTI behaviour in OpenAssetIO exceptions. This was fixed
upstream in OpenAssetIO/OpenAssetIO#1225.

Signed-off-by: David Feltell <david.feltell@foundry.com>

feltech committed:
[Core] Remove public header

Part of #14. Remove unnecessary public header. The resolver plugin is
never `#include`d as an external library, so having a public header just
confuses matters.

Signed-off-by: David Feltell <david.feltell@foundry.com>

feltech committed:
[Core] Full `file://` syntax support

Closes #14. Fully support `file` URLs by making use the OpenAssetIO
`FileUrlPathConverter` utility.

The utility will throw on non-`file` URLs being provided, maintaining
existing behaviour (with a slightly different error message). Confirmed
with manual testing.

The `FileUrlPathConverter` utility is only available as of OpenAssetIO
1.0.0-beta.2.0. So update README.

Signed-off-by: David Feltell <david.feltell@foundry.com>

feltech committed:
[Lint] Minor Clang-Tidy tweaks

Part of #14. A few minor tweaks highlighted by Clang-Tidy - a
few `final` and `const`.

Signed-off-by: David Feltell <david.feltell@foundry.com>
